[13.0][MIG] stock_storage_type

grindtildeath opened the pull request:
Depends on:
 - [x] https://github.com/OCA/server-tools/pull/1751

Includes:
 - [x] https://github.com/OCA/wms/pull/7

grindtildeath committed:
[ADD] stock_storage_type

oca-travis committed:
[UPD] Update stock_storage_type.pot

grindtildeath committed:
Move restrictions from stock_putaway_storage_type_strategy to base module

Define max height/max weight restriction fields

Dedicated tree view to edit package/location storage type relation

Rename fields to drop stock_ prefix

grindtildeath committed:
[IMP] stock_storage_type: black, isort

grindtildeath committed:
[MIG] stock_storage_type: Migration to 13.0

Add tests

Fix dependency issue for pytest

[IMP] pre-commit XML files

Add setuptools

Add module stock_storage_type_putaway_strategy

[IMP] stock_storage_type_putaway_strategy: black, isort

[MIG] stock_storage_type_putaway_strategy: Migration to 13.0

Compute domain on stock.package_level location_dest_id according to storage_type restrictions

Allow to compute restrictions for multiple products and quants

Add constraint on stock.quant according to stock_storage_type

Add tests

Add tests on package destination domain

Add package_level actual location to allowed domain

Fix constraints do not mix products/lots

simahawk reviewed (commented):
LG, please rebase and run pre-commit

grindtildeath commented:
Rebase done.

@simahawk @guewen @jgrandguillaume Here I'm wondering if we shouldn't merge stock_storage_type_putaway_strategy (#13) into this module, because stock_storage_type only defines fields that are unused and don't make any sense by themselves, i.e. without the other module to compute putaways and restrict stock moves according to storage types...

simahawk commented:
@grindtildeath you miss server-tools dependency 
```
odoo.exceptions.UserError: ("You try to install module 'stock_storage_type' that depends on module 'base_m2m_custom_field'.\nBut the latter module is not available in your system.", '')
```

Commit:
Add support of lines without packages in roadmap

grindtildeath committed:
Merge putaway_strategy module into stock_storage_type

Fix failing test and add loggers

Improve error message on stock.quant constraint

jbaudoux commented on stock_storage_type/models/stock_quant.py:
```diff
@@ -0,0 +1,156 @@
+                    ) % (pack_storage_type.name, location.name)
+                )
+            allowed = False
+            package_quants = quant.package_id.mapped('quant_ids')
```
The constrain is evaluated at quant creation. At this stage, the package's quant is not up to date.

```
stock.quant(148,)
(Pdb) self.package_id
stock.quant.package(111,)
(Pdb) self.package_id.quant_ids
stock.quant()
```

jbaudoux commented on stock_storage_type/models/stock_quant.py:
```diff
@@ -0,0 +1,156 @@
+            other_quants_in_location = self.search(
+                [
+                    ('location_id', '=', location.id),
+                    ('id', 'not in', package_quants.ids)
```
We must exclude all previous quants that are not there anymore, i.e. quantity = 0

```suggestion
                    ('id', 'not in', package_quants.ids),
                    ('quantity', '!=', 0)
```

jbaudoux commented on stock_storage_type/models/stock_quant.py:
```diff
@@ -0,0 +1,156 @@
+            loc_storage_types = location.allowed_location_storage_type_ids
+            if (
+                not quant.package_id
+                or not pack_storage_type
```
I would still evaluate the location storage type constrain like do not mix product even if the pack storage type is not defined and even if there is not package. This condition (having a package with a storage type) is only for the "lst_allowed_for_pst" check. For the other checks, we can still apply them based only the location storage type constrain

jbaudoux reviewed (changes requested):
What about extracting the location storage type in a separate module that allows to define restrictions on the location (empty, do not mix) without needing package.

This module would add the allowed storage type concept on the location storage type and manage package.

Commit:
Apply pre-commit

Check waiting moves in "Not Empty" put-away constraints

When an unreserved move's destination is already a bin, we assume:

* the move is probably waiting on another that will be processed soon
* as the destination of the move is already a bin (not Stock, ...), the
  good *will* arrive in this precise location and not another

So we prevent new move lines to target this location.

Note: as we do not know the lot yet in the move, we cannot check if we
have a planned lot for the same lot in the "Do not mix same lot": ignore
it.

gurneyalex committed:
[IMP] stock_storage_type

improve perfs
* in _compute_allowed_location_dest_ids handle the case of a missing pack level, and short cut
* move a mapped() call out of a loop
* use the stored related field state of stock.move.line instead of a dotted path in the search domain

gurneyalex commented:
https://github.com/grindtildeath/wms/pull/5 fixes some performance issues

gurneyalex committed:
Improve put-away location selection perfs

The choice of a suitable stock location was implemented using an
iteration on all candidate locations + several filters running for each.
This is not efficient when we have several 1000s of locations (each
check was taking 30ms -> several minutes required to compute a putaway
location). This commit reimplements the algorithm by using queries to
select locations matching the rules. In order to be able to compute this
using the ORM, we added some technical fields which should not harm
performance too much.

Co-authored-by: Guewen Baconnier <guewen.baconnier@camptocamp.com>
Co-authored-by: Akim Juillerat <akim.juillerat@camptocamp.com>

Commit:
Find leaf sub-locations instead of all children

We will never do a put-away in an intermediate location,
so they should never be kept as candidate locations for the
put-away.

Commit:
Add oca_dependencies.txt

Required by stock_storage_type

Commit:
Replace mapped by a SQL query

Which is much smaller with thousands of locations for a couple of
storage types.

I recycled the field location_ids on the location storage type which
was a unused field (probably forgotten during a refactoring).

Commit:
Use a pre-generated domain to filter allowed locations

When we select a package in a package level, we want to restrict
the selection of the location destination based on the put-away rules.

The previous implementation used a computed M2m field, added in the
view as an invisible field, and a domain on the location_dest_id field
filtering the locations based on this m2m.

The issue: with 5000 allowed locations, changing the package triggers
an onchange (that we want), and the server answers with 5000 ids,
which must be processed by the M2m field client side. It blocks
the browser for several seconds.

The module web_domain_field [0] allow to use a domain stored in a Char
field rather than a field in the XML (read the module's description if
not clear). Using it solves this issue, as the client doesn't need to
evaluate and update a m2m field, it only pass the domain downstream.

[0] https://github.com/OCA/web/tree/13.0/web_domain_field

jbaudoux reviewed (approved):
Thanks for the fixes and perf improvements

Commit:
Remove constrains forbidding 'none' + storage type on location

It may be a valid use case, for instance to be able to send the good on
the "zone" above the bins (with a following new move to go from the Zone
to the bin).

jgrandguillaume reviewed (approved):
Tested & approved

grindtildeath committed:
pre-commit

grindtildeath commented:
I squashed all the commits with a group by authors. Hope this is fine for everyone involved

guewen commented:
> I squashed all the commits with a group by authors. Hope this is fine for everyone involved

Not such a big deal, but a commit message that contains 'do X AND Y', pretty often means you should have 2 separate changes.
For instance, the commit that replaces the XML domain by the field domain using web_domain_field was a good atomic commit that I intended to use as an example for a blog post. Now, if you want to understand, you have to read the whole commit message and reverse engineer what part of the commit's diff is related to what part of the commit's message.

Ref good to keep in mind: https://blog.mocoso.co.uk/talks/2015/01/12/telling-stories-through-your-commits/

Anyway, the test should be fixed by my last fixup, I'll squash it if the tests pass.

guewen commented:
/ocabot merge patch
